### PR TITLE
Get status of VirtualService if available and use that to determine ingress readiness

### LINKED ISSUE #124: Investigate and use Istio 1.6 CRD Status
According to https://istio.io/docs/reference/config/config-status/  Istio CRD Status is now available.

We should investigate and adopt this new feature.

### PR DESCRIPTION
**Pull request by arturenault**
/fixes #124

### CODE REVIEW

**Comment by tcnghia** [comment]
@arturenault it seems istio/istio#28003 is closed. Are we still blocked here?

**Comment by arturenault** [comment]
We were blocked until the release of 1.8.0, which it looks like just happened yesterday. I'll start the process of consuming it.

Once that's in, we can merge this in and start testing it, but won't be able to enable this by default until istio/istio#28797 is fixed.

**Comment by arturenault** [comment]
OK, sorry for the noise. This is finally ready for review @nak3 @ZhiminXiang  :)

**Comment by arturenault** [comment]
@nak3 @JRBANCEL (since zhimin is oof on parental leave), can I get some eyes on this?

File: pkg/reconciler/ingress/ingress.go [hunk 1]
```diff
@@ -213,6 +216,16 @@ func (r *Reconciler) reconcileIngress(ctx context.Context, ing *v1alpha1.Ingress
 		// about the steady state.
 		logger.Debug("Kingress is ready, skipping probe.")
 		ready = true
+	} else if hasStatus, readyStatus := r.virtualServicesReady(ctx, vses); hasStatus {
+		// Check if our VirtualServices have a status property.
```
Comment (JRBANCEL) [inline pkg/reconciler/ingress/ingress.go:220]: Should this check be first? Probing initialization is not cheap.

File: pkg/reconciler/ingress/ingress.go [hunk 2]
```diff
@@ -511,3 +524,74 @@ func isIngressPublic(ing *v1alpha1.Ingress) bool {
 	}
 	return false
 }
+
+// virtualServicesReady checks if a virtual service has a status, and if so if it's ready.
+// the return values are (hasStatus, ready, error), where
+//	hasStatus indicates whether the virtualService has a status field
+//	ready indicates whether it's been reconciled and able to receive requests
+//	error contains any errors that occurred during this
+func (r *Reconciler) virtualServicesReady(ctx context.Context, vses []*v1alpha3.VirtualService) (bool, bool) {
```
Comment (JRBANCEL) [inline pkg/reconciler/ingress/ingress.go:533]: areVirtualServicesReady?

File: pkg/reconciler/ingress/ingress.go [hunk 3]
```diff
@@ -511,3 +524,74 @@ func isIngressPublic(ing *v1alpha1.Ingress) bool {
...
+	var errs []error
+	hasStatus := true
+	ready := true
+	for _, vs := range vses {
+		_hasStatus, _ready, err := r.virtualServiceReady(ctx, vs)
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		hasStatus = hasStatus && _hasStatus
```
Comment (JRBANCEL) [inline pkg/reconciler/ingress/ingress.go:545]: Should we exit the loop if hasStatus is false?

File: pkg/reconciler/ingress/ingress.go [hunk 4]
```diff
@@ -511,3 +524,74 @@ func isIngressPublic(ing *v1alpha1.Ingress) bool {
...
+	if currentState.Generation != currentState.Status.ObservedGeneration {
+		logger.Debugf("VirtualService %v status is stale; checking again...", vs.Name)
+		return true, false, nil
+	}
+
+	var condition *istiov1alpha1.IstioCondition
+	for _, cond := range currentState.Status.Conditions {
+		// Reconciled condition can be "true", "false", or "unknown"
+		if strings.EqualFold(cond.Type, virtualServiceConditionReconciled) {
+			condition = cond
```
Comment (JRBANCEL) [inline pkg/reconciler/ingress/ingress.go:587]: Exit loop here?

File: pkg/reconciler/ingress/ingress.go (hunk 1 above)
Comment (arturenault) [reply pkg/reconciler/ingress/ingress.go:220]: Which probing initialization are you talking about? This is called before statusManager.IsReady

Comment (JRBANCEL) [reply pkg/reconciler/ingress/ingress.go:220]: Ah yes, my bad.
`ing.IsReady()` looks a lot like `statusManager.IsReady(ctx, ing)`

**Review by tcnghia — APPROVED** [review]
/lgtm

@arturenault it may be good for us to have some e2e coverage?

may be do this for istio-latest leg?

**Pull request merged**